src/NHapi.Base/Model/GenericMessage.cs

```diff
@@ -54,31 +57,22 @@ public static Type GetGenericMessageClass(string version)
                 throw new ArgumentException($"The version {version} is not recognized");
             }
 
-            // TODO: Think about converting this to a switch statement
-            Type c = null;
-            if (version.Equals("2.1"))
-            {
-                c = typeof(V21);
-            }
-            else if (version.Equals("2.2"))
-            {
-                c = typeof(V22);
-            }
-            else if (version.Equals("2.3"))
-            {
-                c = typeof(V23);
-            }
-            else if (version.Equals("2.3.1"))
+            Type c;
+            if (GenericMessages.ContainsKey(version))
             {
-                c = typeof(V231);
+                c = GenericMessages[version];
             }
-            else if (version.Equals("2.4"))
+            else
             {
-                c = typeof(V24);
-            }
-            else if (version.Equals("2.5"))
-            {
-                c = typeof(V25);
+                var versionClassName = $"V{version.Replace(".", string.Empty)}";
+                var fullName = $"{typeof(GenericMessage).FullName}+{versionClassName}";
+
+                c = Type.GetType(fullName, false);
+
+                if (c != null)
+                {
+                    GenericMessages.Add(version, c);

```

Comment:
does this need to be thread safe? ie use a `lock` object or concurrent dictionary

Reply:
Concurrent dictionary is not supported in `net3.5` so I'll leave this one for now and add a todo for when we target `netstandard2.0`.